scvi/external/solo/_model.py

```diff
@@ -0,0 +1,162 @@
+import io
+from contextlib import redirect_stdout
+from typing import Optional
+
+import numpy as np
+from anndata import AnnData
+from scipy.sparse import issparse
+
+from scvi import _CONSTANTS
+from scvi.data import get_from_registry, setup_anndata
+from scvi.dataloaders import AnnDataLoader
+from scvi.lightning import ClassifierTrainingPlan
+from scvi.model import SCVI
+from scvi.model.base import BaseModelClass
+from scvi.modules import Classifier
+
+
+LABELS_KEY = "_solo_doub_sim"
+
+
+class SOLO(BaseModelClass):
+    """
+    Reimplementation of Solo for doublet detection [Bernstein19]_.
+
+    Parameters
+    ----------
+    adata
+        single-cell AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
+    latent_obsm_key
+        Key in adata.obsm containing SCVI latent representation
+    use_gpu
+        Use the GPU or not.
+    **classifier_kwargs
+        Keyword args for :class:`~scvi.modules.Classifier`
+
+    Examples
+    --------
+    >>> adata = anndata.read_h5ad(path_to_anndata)
+    >>> scvi.data.setup_anndata(adata)
+    >>> vae = scvi.model.SCVI(adata)
+    >>> vae.train()
+    >>> solo = scvi.external.SOLO.from_scvi_model(vae)
+    >>> solo.train()
+    >>> solo.predict()
+    """
+
```

Comment:
@adamgayoso Thanks again so much for helping me get started in a huge way! Just wanted to put some thoughts down before the long weekend.

Rough sketch of the solo model currently:

1. Train VAE on experimental data (putative singlets)
2. Generate doublets
3. Embed singlet and doublets with VAE
4. Train classifier on those embeddings

This is a very rough sketch of one way to handle steps 2-4 in the SOLO class...at least I think. I'm sure I'm missing some of `scvi-tools` best practices
```
class SOLO(BaseModelClass):
    """
    Reimplementation of Solo for doublet detection [Bernstein19]_.

    Parameters
    ----------
    adata
        single-cell AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
    latent_obsm_key
        Key in adata.obsm containing SCVI latent representation
    use_gpu
        Use the GPU or not.
    **classifier_kwargs
        Keyword args for :class:`~scvi.modules.Classifier`

    Examples
    --------
    >>> adata = anndata.read_h5ad(path_to_anndata)
    >>> scvi.data.setup_anndata(adata)
    >>> vae = scvi.model.SCVI(adata)
    >>> vae.train()
    >>> solo = scvi.external.SOLO.from_scvi_model(vae)
    >>> solo.train()
    >>> solo.predict()
    """

    def __init__(
        self,
        adata: AnnData,
        scvi_model: SCVI,
        use_gpu: bool = True,
        **model_kwargs,
    ):
        self.scvi_model = scvi_model
# I think it was easier to work on adata. directly...
#         x = get_from_registry(adata, _CONSTANTS.X_KEY)
#         if issparse(x):
#             library = np.log(x.sum(1).A.ravel().reshape(-1, 1))
#         else:
#             library = np.log(x.sum(1).reshape(-1, 1))

        self._classifier_count_data = adata
        self._classifier_count_data = self.add_doublets(self._classifier_count_data)
        
        self._classifier_latent_data = self.scvi_model.get_latent_library_size(self._classifier_count_data)
        self._classifier_latent_data = np.concatenate((self._classifier_latent_data, self.scvi_model.get_latent_representation(self._classifier_count_data)), axis=1)
        
        self._classifier_data = AnnData(self._classifier_latent_data, obs=self._classifier_count_data.obs)
        
        f = io.StringIO()
        with redirect_stdout(f):
            setup_anndata(self._classifier_data, labels_key=LABELS_KEY)
        super().__init__(self._classifier_data, use_gpu=use_gpu)
        self.module = Classifier(
            n_input=self._classifier_data.shape[1],
            n_labels=2,
            logits=True,
            **model_kwargs,
        )
        self._model_summary_string = "Solo model"
        self.init_params_ = self._get_init_params(locals())

    @classmethod
    def from_scvi_model(cls, scvi_model: SCVI):
        """Instantiate a model from an scvi model."""
        latent_obsm_key = "_solo_latent_rep"
        _validate_scvi_model(scvi_model)
        latent_rep = scvi_model.get_latent_representation()
        scvi_model.adata.obsm[latent_obsm_key] = latent_rep
        return cls(scvi_model.adata, latent_obsm_key)

    @staticmethod
    def add_doublets(adata: AnnData, 
                     seed: int = 1,
                     doublet_ratio: int = 2):
        def doublet_generating_function(X, i, j):
            return (X[i, :] + X[j, :])
        
        # TODO needs a random seed, use scvi.settings.seed
        scvi.settings.seed = seed
        num_doublets = doublet_ratio * adata.X.shape[0]

        in_silico_doublets = np.zeros((num_doublets, adata.shape[1]))
        # for desired # doublets
        X = adata.X.todense()

        for di in range(num_doublets):
            # sample two cells
            i, j = np.random.choice(adata.shape[0], size=2)
            # generate doublets
            in_silico_doublets[di, :] = doublet_generating_function(X, i, j) 
        
        in_silico_doublets_ad = AnnData(in_silico_doublets)
        
        # add an obs column with name LABELS_KEY of strings, "singlet", "doublet"
        in_silico_doublets_ad.obs.loc[:, LABELS_KEY] = 'doublet'
        adata.obs.loc[:, LABELS_KEY] = 'singlet'
        
        setup_anndata(in_silico_doublets_ad)
        
        adata = adata.concatenate(in_silico_doublets_ad)
        return adata
```

Reply:
@njbernstein, so I'm assuming what you've pasted here works? Like does it train? I can add this doublets function with some minor changes

Reply:
ah I think I messed up the flow of steps in my head. I'll add this doublets function and make things a bit more clear

Reply:
The thing we have to keep in mind is that the init of SOLO cannot have the scvi model, as we pickle all the inputs, and scvi model has it's own save method. So the workaround is to use the class method, where the class method will create the doublets, jointly embed all cells + doublets and then initialize the model, which will take the anndata of cells, doublets and the keys in obsm of latent rep, library size etc.

Reply:
Ah cool. Thank you! I'll dig in more on Tuesday